OpenAPI/kiota/models.md

```diff
@@ -21,6 +21,27 @@ All models declared inline with an operation will be generated under the namespa
 
 Models in an [allOf](https://spec.openapis.org/oas/latest.html#composition-and-inheritance-polymorphism) schema declaration will inherit from each other. Where the uppermost type in the collection is the greatest ancestor of the chain.
 
+### allOf interpretation rules
+
+The following table describes how kiota will project any schema with allOf entries when processing an API description:
+
+| Number of Properties | Number of allOf entries without a reference / inline | Number of allOf entries with a reference | Total number of allOf entries | Result |
+| -------------------- | ---------------------------------------- | ------------------------------------------- | ----------------------------- | ------ |
+| 0 | 0 | 0 | 0 | Ignored/Invalid |
+| 0 | 0 or 1 | 0 or 1 | 1 | Ignored, will process the only allOf entry instead and use the original schema's description and name. |
+| 1 or more | 0 | 0 | 0 | Class/interface without a parent type. |
+| 0 | 1 | 1 | 2 | Class/interface with properties from the inline schema and a parent type from the referenced schema. |
+| 1 or more | 0 | 1 | 1 | Class/interface with properties from the current schema and a parent type from the referenced schema. |
+| 1 or more | 1 | 0 | 1 | Class/interface with properties from the current schema and a parent type from the inline schema. |
+| 0 | 0 | 1 | 1 | Class/interface with properties from the referenced schema and without a parent type. |
+| 0 | 1 | 0 | 1 | Class/interface with properties from the inline schema and without a parent type. |
+| 1 or more | 1 | 1 | 2 | Class/interface with properties from the current schema and a parent type from referenced schema if it has properties. |
+| 1 or more | 1 | 1 | 2 | Class/interface with properties from the current schema and a parent type from the inline schema if it has properties. |
```

Comment:
This is the same condition as the previous line but with a different result.

Reply:
Could we simplify these rules to say an allOf that contains a single $ref causes the current type to inherit from the referenced type and every other schema is merged into the current schema?
Do we believe there is a real scenario for making an inline schema a parent type?

Reply:
> This is the same condition as the previous line but with a different result.

No, we use the inline schema this time instead of the referenced one. This is because sub-conditions have changed (which one has properties)

> Could we simplify these rules to say an allOf that contains a single $ref causes the current type to inherit from the referenced type and every other schema is merged into the current schema? 

We could but we'd be mixing the merging logic with the inheritance once, right now those are two separate sections. I'm not sure what we'd achieve by doing that.

> Do we believe there is a real scenario for making an inline schema a parent type?

Yes, we saw a few descriptions with that case and I've had to implement it + unit tests.

Reply:
This comment should qualify that it assumes the referenced schema has no properties and is therefore ignored.  

We should add another line that covers the case where both inline and referenced schema have properties.

Reply:
```suggestion
| 1 or more | 1 | 1 | 2 | If the inline schema has properties and the referenced schema does not. class/interface with properties from the current schema and a parent type from the inline schema. |
```